File: 2019-2020_programmers/_11_/src/typescript/shuttle_bus_tail_recursion.ts
const parser = str => {
  let arr: string[] = str.split("/").map(s => s.split("|"));
  let [n, m, t] = arr[0];
  let timetable = arr[1];
  return [n, m, t, timetable];
};

//console.log((<any>parser)("1|1|5/08:00|08:01|08:02|08:03"));

let sample = [
  "1|1|5/08:00|08:01|08:02|08:03",
  "1|1|5/08:00|08:01|08:02|08:03|08:04",
  "2|10|2/09:10|09:09|08:00",
  "2|1|2/09:00|09:00|09:00|09:00",
  "1|1|5/00:01|00:01|00:01|00:01|00:01",
  "1|1|1/23:59",
  "10|60|45/23:59|23:59|23:59|23:59|23:59|23:59|23:59|23:59|23:59|23:59|23:59|23:59|23:59|23:59|23:59|23:59"
];

const clock = n =>
  parseInt(n) >= 10 ? n : parseInt(n) >= 0 ? "0" + n : n + 24;
const range = n => Array.from(Array(n).keys());
const getChangedTime = (basetime, num) => {
  const HHMM = [Math.floor(num / 60), num % 60];
  return new Date(basetime.getTime() + num * 60 * 1000);
  //return basetime;
};

const basetime = new Date(0);

const solution = (n: number, t: number, m: number, _timetable: string[]) => {
  let current_time = getChangedTime(new Date(0), 60 * 9);
  let timetable: Date[] = _timetable
    .map(e => {
      const HHMM = e.split(":");
      const depart_time = getChangedTime(
        new Date(0),
        parseInt(HHMM[0], 10) * 60 + parseInt(HHMM[1], 10)
      );
      return depart_time;
    })
    .sort((d1, d2) => d2.getTime() - d1.getTime());

  console.log("---------------------------");
  console.log("n: " + n, ", m: " + m, ", t: " + t);
  console.log(
    "timetable: \n\t[" +
      timetable.map(x =>
        [clock(x.getHours() - basetime.getHours()), clock(x.getMinutes())].join(
          ":"
        )
      ) +
      "]"
  );
  console.log("---------------------------\n");

  const recur = n => {
    console.log(
      "# n: " +
        n +
        ", current_time: " +
        clock(current_time.getHours() - basetime.getHours()) +
        ":" +
        clock(current_time.getMinutes())
    );
    // 정류장에 도착한 인원
    let arrived_crew = timetable.filter(
      x => x.getTime() <= current_time.getTime()
    );
    console.log("  crew " + arrived_crew.length + " arrived ");
    console.log(
      "\t[" +
        arrived_crew.map(x =>
          [
            clock(x.getHours() - basetime.getHours()),
            clock(x.getMinutes())
          ].join(":")
        ) +
        "]"
    );

    // n == 0이면 m(탑승인원) - 1번째 크루의 도착시간
    if (n == 0) {
      if (arrived_crew.length <= m - 1) {
        return current_time;
      } else {
        return getChangedTime(
          arrived_crew[arrived_crew.length - 1 - (m - 1)],
          -1
        );
      }
    }

    // n > 0이면 m(탑승인원)만큼 크루에서 제거하고 recur(--n);
    if (n > 0) {
      for (let i = 0; i < Math.min(arrived_crew.length, m); i++) {
        timetable.pop();
      }
      current_time = getChangedTime(current_time, t);
      return recur(--n);
    }
  }; // recur

  let ret = recur(--n);
  return (
    clock(ret.getHours() - basetime.getHours()) + ":" + clock(ret.getMinutes())
  );
}; // solution

let answer = (<any>solution)(...parser(sample[5]));
console.log(answer);
//let sample2 = [
//{
//n: 1,
//t: 1,
//m: 5,
//timetable: ["08:00", "08:01", "08:02", "08:03"]
//},
//{
//n: 2,
//t: 10,
//m: 2,
//timetable: ["09:10", "09:09", "08:00"]
//},
//{
//n: 2,
//t: 1,
//m: 2,
//timetable: ["09:00", "09:00", "09:00", "09:00"]
//},
//{
//n: 1,
//t: 1,
//m: 5,
//timetable: ["00:01", "00:01", "00:01", "00:01", "00:01"]
//},
//{
//n: 1,
//t: 1,
//m: 1,
//timetable: ["23:59"]
//},
//{
//n: 10,
//t: 60,
//m: 45,
//timetable: [
//"23:59",
//"23:59",
//"23:59",
//"23:59",
//"23:59",
//"23:59",
//"23:59",
//"23:59",
//"23:59",
//"23:59",
//"23:59",
//"23:59",
//"23:59",
//"23:59",
//"23:59",
//"23:59"
//]
//}
//];

File: 2019-2020_programmers/_11_/src/java/ShuttleBusTailRecursion.java

import java.util.Arrays;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.Comparator;

class ShuttleBusTailRecursion {
  int n, t, m; List<LocalTime> timetable; LocalTime currentTime;
  public static void main(String[] args) {
    ShuttleBusTailRecursion bus = new ShuttleBusTailRecursion();
    bus.init(new String[] {
          // "1|1|5/08:00|08:01|08:02|08:03|08:04", 
          "1|1|5/08:00|08:01|08:02|08:03", 
          "2|10|2/09:10|09:09|08:00",
          "2|1|2/09:00|09:00|09:00|09:00",
          "1|1|5/00:01|00:01|00:01|00:01|00:01",
          "1|1|1/23:59",
          "10|60|45/23:59|23:59|23:59|23:59|23:59|23:59|23:59|23:59|23:59|23:59|23:59|23:59|23:59|23:59|23:59|23:59"
    }, 2);
    System.out.println(bus.solve());

    // System.out.println(bus.solution(1, 1, 5, new String[]{
    //   "08:00", "08:01",
    //   "08:02", "08:03",
    //   "08:04" 
    // }));
  }
  // case2: ArrayList<ArrayList<String>>
  // System.out.println(
  // (ArrayList<ArrayList<String>>) arr.stream()
  // .map(x -> new ArrayList<String>(Arrays.asList(x.split("\\|"))))
  // .collect(Collectors.toCollection(ArrayList<ArrayList<String>>::new))
  // );


  public Function<String, Integer> toNum = (s) -> Integer.parseInt(s);

  public void init(String[] __timetable, int i) {
    List<String> arr =  Arrays.asList(__timetable[i].split("/"));
    String[] nmt = arr.get(0).split("\\|");
    String[] _timetable = arr.get(1).split("\\|");
    this.currentTime = LocalTime.parse("09:00", DateTimeFormatter.ofPattern("H:mm"));
    this.n = toNum.apply(nmt[0]);
    this.t = toNum.apply(nmt[1]);
    this.m = toNum.apply(nmt[2]);
    this.timetable = Arrays.asList(_timetable)
      .stream()
      .map(x -> LocalTime.parse(x, DateTimeFormatter.ofPattern("H:mm")))
      .sorted(Comparator.reverseOrder())
      .collect(Collectors.toList());
  }

  Function<Integer, LocalTime> lambdaRecur = (num) -> {
    System.out.println(String.format("# n: %d, current_time: %s\n", n, currentTime));
    List<LocalTime> arrived_crew 
      = timetable.stream()
        .filter(x -> x.isBefore(currentTime) || x.equals(currentTime))
        .collect(Collectors.toList());
    System.out.println(String.format("  crew %s arrived", arrived_crew.size()));
    System.out.println(String.format("\t[ %s ]", 
      arrived_crew.stream()
        .map(x -> x.toString())
        .collect(Collectors.joining(","))
    ));
    if (num == 0) {
        if (arrived_crew.size() <= m - 1) {
            return currentTime;
        } else {
          return arrived_crew.get(arrived_crew.size() - 1 - (m - 1)).minusMinutes(1);
        }
    }
    if (n > 0) {
      IntStream.range(0, Math.min(arrived_crew.size(), m))
        .forEach(x -> timetable.remove(timetable.size() - 1));
      currentTime = currentTime.plusMinutes(t);
      return this.lambdaRecur.apply(--num);
    }
    return this.lambdaRecur.apply(--num);
  };

  @Override
  public String toString() {
    return String.format("n = %d, t = %d, m = %d\ntimetable: \n\t%s", n, t, m, timetable);
  }

  private LocalTime solve() {
    System.out.println("---------------------");
    System.out.println(this.toString());
    System.out.println("---------------------\n");
    return lambdaRecur.apply(--n);
  }

  private LocalTime solution(int n, int t, int m, String[] timetable) {
    this.m = m;
    this.n = n;
    this.t = t;
    this.currentTime = LocalTime.parse("09:00", DateTimeFormatter.ofPattern("H:mm"));
    this.timetable = Arrays.asList(timetable)
      .stream()
      .map(x -> LocalTime.parse(x, DateTimeFormatter.ofPattern("H:mm")))
      .sorted(Comparator.reverseOrder())
      .collect(Collectors.toList());
    System.out.println("---------------------");
    System.out.println(this.toString());
    System.out.println("---------------------\n");
    return lambdaRecur.apply(--this.n);
  }
}
